src/Algebra/Graph.hs

```diff
@@ -1066,3 +1105,73 @@ sparsify graph = res
         m <- get
         put (m + 1)
         overlay <$> s `x` m <*> m `y` t
+
+{- Note [The rules of foldg]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The rules for foldg work like this.
+
+* Up to (but not including) phase 1, we use the "buildG/f" rule to
+  rewrite all saturated applications of f with its buildG/foldg
+  form, hoping for fusion to happen (through the "foldg/buildG" rule).
+
+  In phase 1 and 0, we switch off these rules, inline buildG, and
+  switch on the "graph/f" rule, which rewrites the "foldg/f"
+  thing back into plain map.
+
+  It's important that these two rules aren't both active at once
+  (along with build's unfolding) else we'd get an infinite loop
+  in the rules. Hence the activation control below.
+
+* R functions (R for Rewrite) are here to remember f after applying a
+  "builG/f" rule. These functions are higher-order functions and therefore
+  benefits from inlining.
+
+* The "mapR/mapR" rule optimises compositions of fmapR
+-}
+
+buildG :: forall a. F.Fold a -> Graph a
+buildG = F.foldg Empty Vertex Overlay Connect
+{-# INLINE [1] buildG #-}
+
+-- R functions
+composeR :: (b -> c) -> (a -> b) -> a -> c
+composeR = (.)
+{-# INLINE [0] composeR #-}
+
+matchR :: b -> (a -> b) -> (a -> Bool) -> a -> b
+matchR e v p = \x -> if p x then v x else e
+{-# INLINE [0] matchR #-}
+
+flipR :: (b -> b -> b) -> (b -> b -> b)
+flipR = flip
+{-# INLINE [0] flipR #-}
+
+-- Rules to transform a buildG-equivalent function into its equivalent
+{-# RULES
+-- Transform a fmapR into its build equivalent
+"buildG/map"       [~1] forall f g.
+  fmapR   f g  = buildG (F.Fold $ \e v o c -> foldg e (composeR v f) o c g)
+
+-- Transform a induce into its build equivalent
+"buildG/induce"    [~1] forall p g.
+  induce p g  = buildG (F.Fold $ \e v o c -> foldg e (matchR e v p) o c g)
+
+"buildG/transpose" [~1] forall g.
+  foldg Empty Vertex Overlay (flipR Connect) g = buildG (F.Fold $ \e v o c -> foldg e v o (flip c) g)
```

Comment:
I wonder if it makes sense to make this rule more general and get rid of `flipR` by doing so:

```
"buildG/transpose" [~1] forall g.
  foldg Empty Vertex Overlay (f Connect) g = buildG (F.Fold $ \e v o c -> foldg e v o (f c) g)
```
Will this work?

Reply:
Also, perhaps we could also add a similar general rule for `Overlay`:

```
"buildG/transpose" [~1] forall g.
  foldg Empty Vertex (f Overlay) Connect g = buildG (F.Fold $ \e v o c -> foldg e v (f o) c g)
```

Note that `induce` actually applies function `k` to `Overlay` and `Connect` and could therefore be subject to these rules. Is it possible that by adding these general rules, you can get rid of the special rule for `induce`?

Reply:
Thinking about this again, I think my suggestion won't work because if we use `flip` instead of `flipR` then `flip` will probably inline too early, resulting in `\x y -> Connect y x` or something like this and the rule will no longer match... @nobrakal Could you please try and see what actually happens?

Reply:
Perhaps, we could keep `flipR`, but still generalise the rule. Then it should work, and furthermore it will handle not only `flipR` but other functions too.

Reply:
You made me re-think some parts :) , so mainly:

> get rid of flipR by doing so:

We can write (with some rankntypes hack) what you wrote, but this only remove `flipR` from the rule, not from the `transpose` definition :)

The similar for `overlay` is also acceptable, but not usable for `induce`: we need the full-polymorphism of the "k" function, which we don't have since there is a pattern-match inside.

I think this is solved using `Maybe (NonEmptyGraph)` since it uses `Nothing` in place of `Empty`

Reply:
Note that having a fully polymorphic `k` will remove the need to rewrite `induce` back, and thus will simplify a bit all of this !